### PR TITLE
GH-21: As a developer I want type aliases

### PR DESCRIPTION
**Pull request by mor-n4**
See #21.

### CODE REVIEW

**Comment by mor-n4** [comment]
@mmews-n4 It is not done yet, but you could have a first look.

**Review by mmews-n4 — APPROVED** [review]
wow, great addition to n4js!

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/SubtypeJudgment.java [hunk 1]
```diff
@@ -97,22 +94,28 @@ public Result apply(RuleEnvironment G, TypeArgument leftArg, TypeArgument rightA
...
 		}
 		return firstResult;
 	}
 
-	private Result doApply(RuleEnvironment G, TypeRef left, TypeRef right) {
-		if (left == null || right == null) {
+	private Result doApply(RuleEnvironment G, TypeArgument leftArg, TypeArgument rightArg) {
+		// get rid of wildcards by taking their upper/lower bound
+		final TypeRef leftRef = leftArg instanceof Wildcard ? ts.upperBound(G, leftArg) : (TypeRef) leftArg;
+		final TypeRef rightRef = rightArg instanceof Wildcard ? ts.lowerBound(G, rightArg) : (TypeRef) rightArg;
```
[the remark this answers is not in the document]
Comment (mor-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/SubtypeJudgment.java:120]: as discussed: since a different kind of bound is taken for left and right, bounds have to be recomputed after swapping left/right.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/SubtypeJudgment.java [hunk 2]
```diff
@@ -97,22 +94,28 @@ public Result apply(RuleEnvironment G, TypeArgument leftArg, TypeArgument rightA
...
+		if (leftRef == null || rightRef == null) {
 			return failure();
 		}
 
+		final TypeRef left = replaceAndResolveAlias(G, leftRef);
+		final TypeRef right = replaceAndResolveAlias(G, rightRef);
```
[the remark this answers is not in the document]
Comment (mor-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/SubtypeJudgment.java:123]: same as previous comment

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSFunctionValidator.xtend [hunk 3]
```diff
@@ -496,22 +496,22 @@ class N4JSFunctionValidator extends AbstractN4JSDeclarativeValidator {
...
+		val returnTypeRefRaw = funDef.returnTypeRef;
+		if (returnTypeRefRaw === null) {
 			return;
 		}
-		val returnType = returnTypeRef.declaredType;
-		if (returnType === null || returnType.eIsProxy) {
-			return; // a type reference other than ParameterizedTypeRef or unresolved
-		}
 		val G = funDef.newRuleEnvironment;
+		val returnTypeRefRaw2 = tsh.resolveTypeAliases(G, returnTypeRefRaw); // returnTypeRefRaw was taken directly from AST -> must resolve aliases
```
[the remark this answers is not in the document]
Comment (mor-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSFunctionValidator.xtend:504]: Done.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/TypeAliasProcessor.xtend [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+ * Handles resolution of type aliases during post-processing.
+ */
+@Singleton
+class TypeAliasProcessor extends AbstractProcessor {
+
+	@Inject
+	private TypeSystemHelper tsh;
+
+	def void handleTypeAlias(RuleEnvironment G, EObject obj, ASTMetaInfoCache cache) {
+		// Note: it would be great if we could resolve the declaredTypeRef of TypedElement's
```
[the remark this answers is not in the document]
Comment (mor-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/TypeAliasProcessor.xtend:36]: Done.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/TypeAliasProcessor.xtend [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+		if (defType !== null) {
+			resolveAliasInType(G, defType);
+		}
+	}
+
+	def private void resolveAliasInType(RuleEnvironment G, Type type) {
+		if (type instanceof TypeAlias) {
+			return; // do not resolve the 'actualTypeRef' property in type alias itself
+		}
+		val l = newArrayList; // create list up-front to not confuse tree iterator when replacing nodes!
```
[the remark this answers is not in the document]
Comment (mor-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/TypeAliasProcessor.xtend:57]: Done.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/utils/TypeAliasComputer.xtend [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+	// FIXME API doc
+	def package TypeRef resolveTypeAliasFlat(RuleEnvironment G, TypeRef typeRef) {
+		if (!typeRef.isAliasUnresolved) {
+			return typeRef;
+		}
+
+		var ParameterizedTypeRef originalAliasTypeRef = null;
+		var RecursionGuard<TypeAlias> guard = null;
+		var currTypeRef = typeRef;
+		while (true) {
```
[the remark this answers is not in the document]
Comment (mor-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/utils/TypeAliasComputer.xtend:40]: True. I also had the feeling there should be a more elegant way, but couldn't think of anything better. But I think readability is ok-ish and the impact of this is local to the body of this one method, so no major harm done, I guess.

In any case, my main concern was to avoid recursive invocations, here, to not bloat the stack even more than is done by the type system already.

File: tests/org.eclipse.n4js.xpect.ui.tests/testdata_ui/refactoring/type_variable/RenameRefactoring_rename_type_variable_method.n4js.xt [hunk 7]
```diff
@@ -26,10 +26,9 @@
 class A<T> {
 	public x: T;
 	/* XPECT renameRefactoring at E to 'F' ---
-	public <[E|F]> foo(e: E) {
+	public <[E|F]> foo(e: [E|F]) {
```
[the remark this answers is not in the document]
Comment (mor-n4) [reply tests/org.eclipse.n4js.xpect.ui.tests/testdata_ui/refactoring/type_variable/RenameRefactoring_rename_type_variable_method.n4js.xt:29]: I was surprised, too!
BTW: while working on part 2 I had temporarily re-broken and then re-re-fixed this 😸
It's related to some "getCorrespondingASTNodeForTypeModelElement" code in the reference finder.

**Comment by mor-n4** [comment]
Last commit of part 1: ae4ac7093fa5b9a31bdf75020311b5a4155658dd

**Comment by mor-n4** [comment]
@mmews-n4 Please have another look!

These are the changes since part 1:
https://github.com/eclipse/n4js/compare/ae4ac7093fa5b9a31bdf75020311b5a4155658dd...GH-21

I think a good starting point would be `N4JS.xcore`, then `N4JS.xtext` and maybe `TypeRefProcessor.xtend`.

**Comment by mor-n4** [comment]
@mmews-n4 FYI, here he remaining items on my TODO list:
- more API doc
- validation for use of alias after extends/implements in class/interface declarations
- consider performance tweak: use flag to denote cyclic type aliases and use that in validation

File: tests/org.eclipse.n4js.spec.tests/xpect-tests/Ch11_01__Array_and_Object_Destructuring/destructuring/Destruct_ExpectedType_SpecialCases.n4js.xt [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+	IssueConfiguration {
+		IssueCode "CFG_USED_BEFORE_DECLARED" { enabled = false }
+	}
+
+END_SETUP */
+
+
+//
+// some challenging special cases for method DestructureHelper#createTypeFromLeafDestructNode()
```
Comment (mmews-n4) [inline tests/org.eclipse.n4js.spec.tests/xpect-tests/Ch11_01__Array_and_Object_Destructuring/destructuring/Destruct_ExpectedType_SpecialCases.n4js.xt:22]: Is this test class related to this userstory?

**Review by mmews-n4 — APPROVED** [review]
Nice to also have really good new features while doing all the refactorings/clean-up stories.

**Comment by mor-n4** [comment]
For reference:

Compared time performance with master and space performance with a version without aggressive copying of type references in `TypeRefProcessor` and both seems to be fine.

The aggressive type-ref copying costs about 0.5% memory at the time of peak memory usage:
```
with copying processed TypeRefs:
MAX USED MEMORY: 1489763 kB
without copying processed TypeRefs:
MAX USED MEMORY: 1483051 kB
```
(measured during full build, after each project before emptying the resource set)

However, the impact is only that large if many resources are loaded from source and then retained (as during a full build). During incremental build and in editors the impact will be much smaller.

Thus, I leave the aggressive copying of type refs in `TypeRefProcessor` active, for now.

**Comment by mor-n4** [comment]
This PR got messed up by a force push on master.
Closing this; continued after rebase with same content as PR #2084.
